### DISCUSSION TITLE
When should I wrap the InboundProtocol in ReadyUpgrade?

### DISCUSSION THREAD

**Discussion by pieterdd** [libp2p/rust-libp2p#3963] [Q&A] [answered]
I have the impression that there are two common ways to implement `ConnectionHandler::listen_protocol`.

On the one hand, here's a [snippet](https://github.com/mxinden/libp2p-perf/blob/c23a7cc485d4dd84c33fbbe4634cc46409b47501/rust/src/handler.rs#L242) from libp2p-perf. Its InboundProtocol type is set to `PerfProtocolConfig`, which implements `InboundUpgrade`.

```rust
fn listen_protocol(&self) -> SubstreamProtocol<Self::InboundProtocol, Self::InboundOpenInfo> {
    SubstreamProtocol::new(PerfProtocolConfig {}, ())
}
```

On the other hand, here's a [snippet](https://github.com/libp2p/rust-libp2p/blob/a2ad916bc4d071dcda6d67483f9c8e798f4195c1/protocols/ping/src/handler.rs#L241) from rust-libp2p's ConnectionHandler for ping. Its InboundProtocol type is set to `ReadyUpgrade<StreamProtocol>`, and `StreamProtocol` does not implement `InboundUpgrade`.

```rust
fn listen_protocol(&self) -> SubstreamProtocol<ReadyUpgrade<StreamProtocol>, ()> {
    SubstreamProtocol::new(ReadyUpgrade::new(PROTOCOL_NAME), ())
}
```

What are the implications of choosing one approach over the other?

**Comment by thomaseizinger** [comment, marked as answer]
There is ongoing work to deprecate and eventually remove the former, see https://github.com/libp2p/rust-libp2p/issues/2863.

Hopefully, a `ConnectionHandler` will in the future just return a list of `StreamProtocols` that it wants to listen on.

In regards to the implications:

The former gives you a bit more type-safety because the connection task will run the upgrade future for you and only give you back the result. The downside is that it is quite verbose and hard to understand because there are many moving pieces:

 - `InboundUpgrade`
 - `OutboundUpgrade`
 - `DialUpgradeError`
 - `ListenUpgradeError`
 - `UpgradeInfo`

Another downside is that there is less flexibility in what you can do with the stream. You have to process it fully within the upgrade traits.

This is where the 2nd approach comes in. By just passing through the stream to the handler, you can do whatever you want with it.

My vision is to offer many, well-composable pieces that make stream handling itself easy and reduce the API surface of `ConnectionHandler`.

Here is some draft work in that direction: https://github.com/mxinden/asynchronous-codec/pull/5

**Reply by pieterdd** [reply]
That clears things up. Thanks Thomas!
